### ISSUE TITLE
imfile: multiline buffer corruption

### ISSUE THREAD

**Issue by mhyllander** [rsyslog/rsyslog#874]
I'm testing a setup where I collect logs using rsyslog imfile and pass them to kafka. Some of the logs (~30%) are multiline. I'm seeing a problem where around 1% of the multiline logs are corrupted. The symptoms are that the first line of the multiline log message is intact, but the rest of the message is corrupt. Initially I discovered the problem because logstash was unable to parse the corrupt logs after retrieving them from kafka. I've tried to isolate the problem by saving the log messages to a local file using omfile, on the source host. The problem appears there as well, so I believe we can safely leave kafka and logstash out of the picture.

**Example of good log message**
Original log lines:
```
1457562691510 [E(6)] INFO  LookupClient  - UPDATE api-test.jobrstreaming.com/192.168.185.222:443 sLen:256 cLen:52
ConnectivityInfoUpdate(guid:2B 24 97 74 F5 EB D2 AE B4 1B 1C 01 A3 36 24 9F, gatewayType:2, publicAddress:InetAddress(address:C0 4B B2 81, port:8001), privateAddress:InetAddress(address:C0 4B B2 81, port:8001), isDialup:false, uploadBandwidth:0, justRelay:true, externalAS:8075, internalAS:0)
```
Log message after collection by imfile and output with omfile using a json template (I've indented the json structure to make it easier to read):
```
{"@timestamp":"2016-03-09T22:34:41.367305+00:00",
  "host":"jobr-test-euw-relay-ry1-0",
  "type":"relay-service",
  "file":"\/opt\/relay-service\/current\/logs\/main.log",
  "message":"1457562691510 [E(6)] INFO  LookupClient  - UPDATE api-test.jobrstreaming.com\/192.168.185.222:443 sLen:256 cLen:52ConnectivityInfoUpdate(guid:2B 24 97 74 F5 EB D2 AE B4 1B 1C 01 A3 36 24 9F, gatewayType:2, publicAddress:InetAddress(address:C0 4B B2 81, port:8001), privateAddress:InetAddress(address:C0 4B B2 81, port:8001), isDialup:false, uploadBandwidth:0, justRelay:true, externalAS:8075, internalAS:0)\\n"
}
```
_(The newline marker `\\n` is misplaced, but I've reported that in a separate issue #843. It should have been placed before the second line instead of after, i.e. between `cLen:52` and `ConnectivityInfoUpdate`.)_

**Example of corrupt log message**
Original log lines:
```
1457569891545 [E(7)] INFO  LookupClient  - UPDATE api-test.jobrstreaming.com/192.168.185.222:443 sLen:256 cLen:52
ConnectivityInfoUpdate(guid:2B 24 97 74 F5 EB D2 AE B4 1B 1C 01 A3 36 24 9F, gatewayType:2, publicAddress:InetAddress(address:C0 4B B2 81, port:8001), privateAddress:InetAddress(address:C0 4B B2 81, port:8001), isDialup:false, uploadBandwidth:0, justRelay:true, externalAS:8075, internalAS:0)
```
Output log message:
```
{"@timestamp":"2016-03-10T00:31:31.548801+00:00",
  "host":"jobr-test-euw-relay-ry1-0",
  "type":"relay-service",
  "file":"\/opt\/relay-service\/current\/logs\/main.log",
  "message":"1457569891545 [E(7)] INFO  LookupClient  - UPDATE api-test.jobrstreaming.com\/192.168.185.222:443 sLen:256 cLen:52\u0000\u0000\u0000\u0000\u0000\u0000\u0000q\u0001\u0000\u0000\u0000\u0000\u0000\u0000 \u0001\u0000D\u0013\u0000\u0000 \u0001\u0000D\u0013\u0000\u0000(7)] DEBUG WebClient  - LOOKUP REQ: https:\/\/api-test.jobrstreaming.com:443\/v1\/lookup\/2b249774f5ebd2aeb41b1c01a33X\u0001\u0000D\u0013\u0000\u0000 \u0000\u0000\u0000\u0000\u0000\u0000\u0000H\u0001\u0000D\u0013\u0000\u0000H\u0001\u0000D\u0013\u0000\u0000pClient  - UPDATE api-test.jobrstreaming.com\/192.168.185.222:443 sLen:256 cLen:52\u00002aeb41b1c01a33 \u0000\u0000\u0000\u0000\u0000\u0000\u0000d\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0010 \u0001D\u0013\u0000"
}
```
As you can see, everything after the first line of the message (after `cLen:52`) is corrupt. This is true for all corrupted log messages. You can also see bits and pieces of other log lines in the corrupt data. (In one case I could see the beginning of the next log line from the file in the middle of the corrupt data.)

As I mentioned, this happens in about 1% of the multiline log messages. Single-line log messages are unaffected. I ran the system for a couple of days and got 28 corrupted log messages out of a total of 2980 multiline log messages.

I'm running this on Ubuntu 14.04.1, using rsyslog (and rsyslog-kafka) version 8.16.0-0adiscon1trusty1 from the http://ppa.launchpad.net/adiscon/v8-stable/ubuntu repository. Below is the configuration I used:

```
module(load="imfile")

main_queue(
  queue.workerthreads="1"      # threads to work on the queue
  queue.dequeueBatchSize="100" # max number of messages to process at once
  queue.size="10000"           # max queue size
)

template(name="debug_json" type="list") {
  constant(value="{")
  constant(value="\"@timestamp\":\"")
  property(name="timereported" dateFormat="rfc3339" format="json")
  constant(value="\",\"host\":\"")
  property(name="hostname" format="json")
  constant(value="\",\"type\":\"")
  property(name="syslogtag" format="json")
  constant(value="\",\"file\":\"")
  property(name="$!metadata!filename" format="json")
  constant(value="\",\"message\":\"")
  property(name="msg" format="json")
  constant(value="\"}\n")
}

ruleset(name="kafka_logstash_debug") {
  action(
    type="omfile"
    file="/mnt/relay.log"
    template="debug_json"
  )
  stop
}

$RepeatedMsgReduction off
input(type="imfile"
  file="/opt/relay-service/current/logs/main.log"
  tag="relay-service"
  PersistStateInterval="100"
  startmsg.regex="^[0-9]{13} "
  addMetaData="on"
  ruleset="kafka_logstash_debug"
)
$RepeatedMsgReduction on
```

**Comment by davidelang** [comment]
There are known corruption issues with 8.16 due to the json-c library not being thread safe. This library has been forked to libfastjson and many (but not necessarily all) holes have been plugged. 8.18 now requires libfastjson, and you can use it with 8.16 if you compile your own version.

could you please test something with libfastjson and see if you still have problems?

**Comment by mhyllander** [comment]
Hi, Thanks for that suggestion. I have been running 8.17 for a while, and yesterday deployed 8.18. Unfortunately I still see these problems. Here's an example from today, an error message from logstash when it can't parse the corrupted message:

```
{:timestamp=>"2016-04-21T13:49:51.222000+0000", :message=>"Received an event that has a different character enc
oding than you configured.", :text=>"{\\\"@timestamp\\\":\\\"2016-04-21T13:49:50.713835+00:00\\\",\\\"host\\\":
\\\"jobr-test-euw-procserv-ps1-1\\\",\\\"type\\\":\\\"jobr-processing\\\",\\\"file\\\":\\\"\\\\/opt\\\\/procser
v\\\\/logs\\\\/jobr-processing.log\\\",\\\"message\\\":\\\"2016-04-21T13:49:50.712 [ERROR] org.hbase.async.Regi
onClient - Unexpected exception from downstream on [id: 0xa1d63b70]\\\\u0000\\x85\\\\u0000\\\\u0000\\\\u0000\\\
\u0000\\\\u0000\\\\u0000\\\\u00002016-04-21T13:49:50.711 [WARN ] org.hbase.async.HBaseClient - Couldn't connect
 to the RegionServer @ 10.0.8.58:60020\\\\u0000\\\\u0000\\\\u0000\\\\u0000!\\\\u0001\\\\u0000\\\\u0000\\\\u0000
\\\\u0000\\\\u0000\\\\u0000\\x88\\\\u0001\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\x88\\\\u0001\\\\u0000xk\\u007F\\
\\u0000\\\\u0000\\x90I\\xFE\\x93k\\u007F\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000Pu\\\\u0000xk\\u007F
\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\
\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0
000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000
\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\
\u0000p\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\x84\\\\u0000\\\\u0000\\\\u0000\\\\u0000
\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000021T13:49:50.
711 [WARN ] org.hbase.async.HBaseClient - Couldn't connect to the RegionServer @ 10.0.8.58:60020, re1\\\\u0000\
\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\x98\\\\u0000\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\x98\\\
\u0000\\\\u0000xk\\u007F\\\\u0000\\\\u0000Pt\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\x88\\\\u0000\\\\u0000xk\\u007
F\\\\u0000\\\\u0000 \\\\u0001\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u00004\\\\u0000\\\\u0000\\\\u0000
\\\\u0000\\\\u0000\\\\u0000\\\\u0000pI\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\
\\\u0000\\\\u0000\\\\u0000\\\\u0000\\x80\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u001
6\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
\\u0000\\\\u0000\\\\u0000\\xF1\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000X\\\\u0002\\\\u000
0xk\\u007F\\\\u0000\\\\u0000X\\\\u0002\\\\u0000xk\\u007F\\\\u0000\\\\u0000:50.713 [INFO ] org.hbase.async.HBase
Client - Added client for region RegionInfo(table=\\\\\\\"counter_set\\\\\\\", region_name=\\\\\\\"counter_set,
\\\\\\\\x08(9\\\\\\\\xD8m\\\\\\\\xBB{\\\\/\\\\\\\\x00\\\\\\\\x00\\\\\\\\x00\\\\\\\\x00T+D\\\\\\\\x00QoE=1,14177
69017272.c0f0ffa34a6533dc4508b270faa661c3.\\\\\\\", stop_key=[15, 8, 110, 37, 76, -11, 18, -123, 0, 0, 0, 0, 84
, 0, 60, 96]), which was updated in the regions cache.  Now we know that RegionClient@786401176(chan=null, #pen
ding_rpcs=0, #batched=0, #rpcs_inflight=0) is hosting 1 region.\\\\u0000\\\\u0000.\\\\u00000_71\\xF0\\\\u0001\\
\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u00004\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\
u0000\\xC0\\xDC\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0
000\\\\u0000\\x80\\\\u0003\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\xC9\\\\u0001\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
x85\\\\u0002\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\xF0;\\\\u0014\\\\u0001\\\\u0000\\\\u0000\\\
\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0002\\\\u0000\\\\u0
000\\\\u0000\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\x80T\\\\u0000xk\\u007F\\\\u0000\\\\u0000&\\\\u0000\\\\u0000\\
\\u0000\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\x80\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000
\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\
\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000FG\\\\
u000B\\\\u0002\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u00
00\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0010\\\\u0000\
\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\
u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0002\\\\u0000\\\\u0000\\\\u0000\\xFF\\xFF\\xFF\\xFF\\\\u0000\\\\u
0000\\\\u0000\\\\u0000\\xE8\\\\u0000\\\\f\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\x90V\\\\u0000xk\\u007F\
\\\u0000\\\\u0000\\\\u0000\\x99\\\\u0000xk\\u007F\\\\u0000\\\\u0000\\xD9\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\\
\u0000\\\\u0000\\\\u0000\\xD9\\\\u0001\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\xFF\\xFF\\xFF\\xF
F\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u
0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u
0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u
0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\
\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u
0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u000
0\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\\"}", :expected_charset=>"U
TF-8", :level=>:warn}
```

**Comment by davidelang** [comment]
I know this is hard to do when you have a small percentage of logs giving you problems, but can you try to get a RSYSLOG_DebugFormat output of the problem messages?

If you can detect when something is malformed, or even a common pattern in the malformed message, you can try filtering on that rather than dumping everything in that verbose format.

I'm very interested to see what rawmsg looks like. In the json-c corruption cases, things under $! or !. will be corrupt, but the $msg and $rawmsg values will be fine. It would be good to see exactly where things are going wrong (and since this is imfile, if you can identify what the section in the file is that is being mis-interpreted it may help as well)

**Comment by mhyllander** [comment]
They corrupt log events occur often enough, so today I have captured with RSYSLOG_DebugFormat. Attached is one example, with the original log lines and the debug output.
[mbc_logs.txt](https://github.com/rsyslog/rsyslog/files/232557/mbc_logs.txt)

**Comment by davidelang** [comment]
Ok, that is showing corruption in the data that json-c doesn't touch

Since this only happens in multi-line logs with you using the regex to detect when a log starts, I'm going to guess it's a problem there. Someone posted a possible issue in that code within the last couple of days.

**Comment by mhyllander** [comment]
I tracked down the problem, please see my pull request.

**Comment by rgerhards** [comment]
fixed via #1042
